draft-irtf-cfrg-hash-to-curve.md

```diff
@@ -2712,28 +2710,121 @@ of the counter byte I2OSP(i, 1)).
+({{hashtofield-expand-xof}}), but the second method MUST NOT be
+used with expand\_message\_xof.
+
+1. For each invocation of H outside hash\_to\_field, choose a unique domain
+   separation tag DST\_ext.
+   Prepend DST\_ext to the input to H, and in addition append a single
+   zero byte to this input.
+
+   Appending the byte I2OSP(0, 1) to all inputs to H outside hash\_to\_field
+   ensures that these inputs are distinct from those inside hash\_to\_field
+   because the final byte of DST\_prime encodes the length of DST, which is
+   required to be nonzero ({{domain-separation}}), and DST\_prime is always
+   appended to invocations of H inside hash\_to\_field.
+   Further, distinct DST\_ext values ensure that invocations of H outside
+   hash\_to\_field are distinct from one another.
+
+   For example, for two invocations of H whose inputs are msg1 and msg2,
+   one might choose distinct DST\_ext1 and DST\_ext2 and compute
+
+        hash1 = H(DST_ext1 || msg1 || I2OSP(0, 1))
+        hash2 = H(DST_ext2 || msg2 || I2OSP(0, 1))
+
+2. (Note: this method MUST NOT be used with expand\_message\_xof.)
+   For each invocation of H outside expand\_message\_xmd, choose a unique domain
+   separation tag DST\_ext.
+   DST\_ext MUST be at least b bits long, where b is the number of bits output
+   by the hash function H, and it MUST contain at least one nonzero bit among
+   the first b bits.
+   Prepend the value DST\_ext to the input to H.
+
+   Since DST\_ext contains at least one nonzero bit among its first b bits,
+   it is guaranteed to be distinct from the value Z\_pad
+   ({{hashtofield-expand-xmd}}, step 4), which ensures that all inputs to H
+   are distinct from the input used to generate b\_0 in expand\_message\_xmd.
+   Moreover, since DST\_ext is at least b bits long, it is almost certainly
+   distinct from the values b\_0 and strxor(b\_0, b\_(i - 1)), and therefore
+   all inputs to H are distinct from the inputs used to generate b\_i, i >= 1,
+   with high probability.
+
+   For example, for two invocations of H whose inputs are msg1 and msg2,
+   one might choose distinct nonzero DST\_ext1 and DST\_ext2 of length
+   b bits each and compute
+
+        hash1 = H(DST_ext1 || msg1)
+        hash2 = H(DST_ext2 || msg2)
+
+Other expand\_message variants that follow the guidelines in
+{{hashtofield-expand-other}} are expected to have similar properties,
+but these should be analyzed on a case-by-case basis.
+
+In addition to the above considerations, protocols that use Merkle-Damgaard
+hash functions as random oracles should be extremely careful to defend
+against length extension and other well known attacks.
+expand\_message\_xmd is designed to guard against these attacks.
+It can be used as a random oracle outside of hash\_to\_field, and
+can be domain separated from invocations inside hash\_to\_field simply
+by choosing a different DST.
+
+Another possibility is to use HMAC {{RFC2104}}.
```

Comment:
This is another possibility to use expand_message_xmd outside of hash_to_field? I'm not sure I follow this example. The above paragraph says, as I understand it, "feel free to crib expand_message_xmd for your purposes so long as you apply domain separation with a separate tag." It's not clear to me how this example follows. Can you clarify?

Reply:
Ah, yes, sorry. What I meant was, an alternative to using expand_msg_xmd is to use HMAC. And then the idea is, this is how you'd domain separate HMAC from expand_message_xmd. I will clarify

Reply:
OK, I took a crack at clarifying. Does this make sense now?

The reason I want to include "how to domain separate HMAC from expand_message_xmd" is that I suspect this might be something that protocols will want to do.

Reply:
Did some further simplification.

Reply:
Yep! All good now.